tests/testhelpers.py

```diff
@@ -1,257 +1,209 @@
+    return False
+
+  for k in keys_src:
+    ent_src = source_obs[k]
+    ent_tgt = target_obs[k]
+    if list(ent_src.keys()) != list(ent_tgt.keys()):
+      if debug:
+        print("entries don't match. key:", k)
+      return False
+
+    obj = ent_src.keys()
+    for o in obj:
+      obj_src = ent_src[o]
+      obj_tgt = ent_tgt[o]
+      if np.sum(obj_src != obj_tgt) > 0:
+        if debug:
+          print("objects don't match. key:", k, ', obj:', o)
+        return False
+
+  return True
+
+
+def player_total(env):
+  sum_gold = 0
+
+  for ent in env.realm.players.values():
+    sum_gold += ent.gold.val
+
+  return sum_gold 
+
+
+def count_actions(tick, actions):
+  cnt_move = 0
+  cnt_attack = 0
+  cnt_sell = 0
+  cnt_use = 0
+  cnt_give = 0
+  cnt_buy = 0
+
+  for entID in list(actions.keys()):
+    for atn, args in actions[entID].items():
+      if atn == nmmo.action.Move:
+        cnt_move += 1
+      elif atn == nmmo.action.Attack:
+        cnt_attack += 1
+      elif atn == nmmo.action.Sell:
+        cnt_sell += 1
+      elif atn == nmmo.action.Use:
+        cnt_use += 1
+      elif atn == nmmo.action.Give:
+        cnt_give += 1
+      elif atn == nmmo.action.Buy:
+        cnt_buy += 1
+      else:
+        print('not counted:', atn)
+
+  print('Tick:', tick, ', alive agents:', len(actions.keys()), 
+        ', atn cnts move:', cnt_move, ', attack:', cnt_attack, ', sell:', cnt_sell, 
+        ', use:', cnt_use, ', give:', cnt_give, ', buy:', cnt_buy)
+
+  return cnt_move, cnt_attack, cnt_sell, cnt_use, cnt_give, cnt_buy
+
+
+class ScriptedAgentTestConfig(nmmo.config.Small, nmmo.config.AllGameSystems):
+
+  __test__ = False
+
+  LOG_ENV = True
+
+  LOG_MILESTONES = True
+  LOG_EVENTS = False # TODO: LOG_EVENTS needs to be fixed
+  LOG_VERBOSE = False
+
+  SPECIALIZE = True
+  PLAYERS = [
+          baselines.Fisher, baselines.Herbalist, baselines.Prospector, baselines.Carver, baselines.Alchemist,
+          baselines.Melee, baselines.Range, baselines.Mage]
+
+
+class ScriptedAgentTestEnv(nmmo.Env):
+  '''
+  EnvTest step() bypasses some differential treatments for scripted agents
+  To do so, actions of scripted must be serialized using the serialize_actions function above
+  '''
+  __test__ = False
+
+  def __init__(self, config: nmmo.config.Config, seed=None):
+    super().__init__(config=config, seed=seed)
+    # this is to cache the actions generated by scripted policies
+    self.actions = {}
+
+  def reset(self, map_id=None, seed=None, options=None):
+    self.actions = {}
+    return super().reset(map_id=map_id, seed=seed, options=options)
+
+  def step(self, actions):
+    assert self.obs is not None, 'step() called before reset'
+
+    # all agent must be scripted agents
+    for ent in self.realm.players.values():
+      assert isinstance(ent.agent, baselines.Scripted) == True, 'All agent must be scripted.'
+
+    # if actions are not provided, generate actions using the scripted policy
+    if actions == {}:
```

Comment:
could we just use _compute_actions here?

Reply:
This part is adapted from the scripted agents' action generation and deserialize in the main Env.step(), which comes after _process_actions. So, the actions generated by scripted agents are not filtered and used as is. 

The part that looks like filtering (`if arg == nmmo.io.action.Price and type(val) != int:`) is to convert discrete values into int and make the actions pickle-able.